Draft proposal for held token standard (#4987)
ethereum/EIPs: EIPS/eip-draft_held_token.md

---

> eip: <to be assigned>

Reviewer: ```suggestion
eip: 4987
```

title: Held token standard
description: Standard interface to query ownership and balance of held tokens
author: Devin Conley (@devinaconley)
discussions-to: https://ethereum-magicians.org/t/erc-standard-for-held-non-fungible-token-nfts-defi/7117
status: Draft
type: Standards Track

> category (*only required for Standards Track): ERC

Reviewer: ```suggestion
category: ERC
```

created: 2021-09-21
requires (*optional): 20, 165, 721, 1155
---

## Abstract

> The proposed standard defines a lightweight interface to expose functional ownership and balances of held tokens. This standard is implemented by smart contracts which hold ERC20, ERC721, or ERC1155 and is intended to be consumed by both on-chain and off-chain systems that rely on ownership and balance verification.

Reviewer: Little extra definition, and links to other EIPs are required the first time they're mentioned:

```suggestion
The proposed standard defines a lightweight interface to expose functional ownership and balances of held tokens. A held token is a token owned by a contract. This standard may be implemented by smart contracts which hold [ERC-20](./eip-20.md), [ERC-721](./eip-721.md), or (ERC-1155)[./eip-1155.md] and is intended to be consumed by both on-chain and off-chain systems that rely on ownership and balance verification.
```

## Motivation

As different areas of crypto (DeFi, NFTs, etc.) converge and composability improves, there will more commonly be a distinction between the actual owner (likely a contract) and the functional owner (likely a user) of a token. Currently, this results in a conflict between mechanisms that require token deposits and systems that rely on those tokens for ownership or balance verification.

This proposal aims to address that conflict by providing a standard interface for token holders to expose ownership and balance information. This will allow users to participate in these DeFi mechanisms without giving up existing token utility. Overall, this would greatly increase interoperability across systems, benefiting both users and protocol developers.

Example implementers of this ERC standard include
- staking or farming contracts
- lending pools
- time lock or vesting vaults
- fractionalized NFT contracts

Example consumers of this ERC standard include
- governance systems
- gaming
- PFP verification
- art galleries or showcases
- token based membership programs

## Specification

> Smart contracts implementing the held token standard MUST also implement `ERC165` and return true when the corresponding interface ID is passed.

Reviewer: You should specify the ERC-165 identifier number explicitly.

Author: fixed

Smart contracts implementing the `ERC20` held token standard MUST implement all of the functions in the `IERC20Holder` interface.

```solidity
/**
 * @notice the ERC20 holder standard provides a common interface to query
 * token balance information
 */
interface IERC20Holder is IERC165 {
  /**

>    * @notice emitted when the token is transferred to the contract

Reviewer: Is this possible to do with ERC20 tokens? I didn't think they called anything when you transfer to a contract, unlike the safe transfer functions of ERC721 and ERC1155.

Author: In the case of ERC20 tokens, the deposit to a holder would need to use some explicit contract function anyways to handle bookkeeping. The emit could be emitted from there

Author: But I think this brings up a bigger point of whether this event should be emitted at all... bringing over to the discussion thread

   * @param owner functional token owner
   * @param tokenAddress held token address
   * @param tokenAmount held token amount
   */
  event Hold(
    address indexed owner,
    address indexed tokenAddress,
    uint256 tokenAmount
  );

  /**
   * @notice emitted when the token is released back to the user
   * @param owner functional token owner
   * @param tokenAddress held token address
   * @param tokenAmount held token amount
   */
  event Release(
    address indexed owner,
    address indexed tokenAddress,
    uint256 tokenAmount
  );

  /**
   * @notice get the held balance of the token owner
   * @param tokenAddress held token address
   * @param owner functional token owner
   * @return held token balance
   */

>   function heldBalanceOf(address tokenAddress, address owner)

Reviewer: Bikeshedding, but maybe `heldFor` would be another name?

    external
    view
    returns (uint256);
}

```

Smart contracts implementing the `ERC721` held token standard MUST implement all of the functions in the `IERC721Holder` interface.

```solidity
/**
 * @notice the ERC721 holder standard provides a common interface to query
 * token ownership and balance information
 */
interface IERC721Holder is IERC165 {
  /**

>    * @notice emitted when the token is transferred to the contract

Reviewer: Similarly, if a token is transferred with `transferFrom` (instead of `safeTransferFrom`), is the implementer expected to raise the event?

Author: Same answer as above, if an ERC721 is sent with the raw `transferFrom` method, the holder contract won't know what to do with it anyways. So this event could be emitted during bookkeeping in either the transfer callback or an explicit deposit method

The token address parameter is included to support contracts that can hold multiple token contracts simultaneously. While some contracts may only hold a single token address, this is more general to either scenario.

Separate interfaces are proposed for each token type (ERC20, ERC721, ERC1155) because any contract logic to support holding these different tokens is likely independent. In the scenario where a single contract does hold multiple token types, it can simply implement each appropriate held token interface.

## Backwards Compatibility

Importantly, the proposed specification is fully compatible with all existing ERC20, ERC721, and ERC1155 token contracts. 

Token holder contracts will need to be updated to implement this lightweight interface.

Consumer of this standard will need to be updated to respect this interface in any relevant ownership logic.

## Reference Implementation

A full example implementation can be found here:

> https://github.com/devinaconley/token-hold-example

Reviewer: Please replace this external link with a _minimal_ reference implementation under `../assets/eip-4987`.

Author: fixed
